plugin/acl/acl.go

```diff
@@ -0,0 +1,103 @@
+package acl
+
+import (
+	"context"
+	"fmt"
+	"net"
+
+	"github.com/coredns/coredns/plugin"
+	"github.com/coredns/coredns/plugin/acl/filter"
+	"github.com/coredns/coredns/plugin/metrics"
+	"github.com/coredns/coredns/request"
+	"github.com/miekg/dns"
+)
+
+type acl struct {
+	Next plugin.Handler
+
+	Rules []Rule
+}
+
+// Rule defines a list of Zones and some ACL policies which will be
+// enforced on them.
+type Rule struct {
+	Zones    []string
+	Policies []Policy
+}
+
+// Policy defines the ACL policy for DNS queries.
+// A policy performs the specified action (block/allow) on all DNS queries
+// matched by source IP or QTYPE.
+type Policy struct {
+	action string
+	qtype  uint16
+	filter filter.Filter
+}
+
+const (
+	// ALLOW allows authorized queries to recurse.
+	ALLOW string = "allow"
+	// BLOCK blocks unauthorized queries towards protected DNS zones.
+	BLOCK string = "block"
+)
+
+func (a acl) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns.Msg) (int, error) {
+	state := request.Request{W: w, Req: r}
+	for _, rule := range a.Rules {
+		// check zone.
+		zone := plugin.Zones(rule.Zones).Matches(state.Name())
```

Comment:
this should be done before the loop

Reply:
> this should be done before the loop

Sorry I don't get it. I'm checking whether the target domain name is included by the zones specified in a single rule.
For instance, the configuration below only works for `example.org`. 
```
. {
  acl example.org {
    allow net 192.168.1.0/24
    block
  }
} 
```